File: packages/python/trikhub/gateway/registry_provider.py
"""
Gateway Registry Provider — implements TrikRegistryContext.

Provides search, install, uninstall, upgrade, list, and getInfo operations
by proxying to the TrikHub registry API and the local gateway.

Mirrors packages/js/gateway/src/registry-provider.ts.
"""

from __future__ import annotations

import asyncio
import json
import shutil
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Protocol

import httpx

from trikhub.manifest.types import (
    InstalledTrikInfo,
    TrikDetailInfo,
    TrikInstallResult,
    TrikSearchResult,
    TrikSearchResultItem,
    TrikUninstallResult,
    TrikUpgradeResult,
)


class RegistryProviderGateway(Protocol):
    """Interface for gateway operations needed by the registry provider."""

    def get_loaded_triks(self) -> dict[str, Any]: ...
    async def load_trik(self, path: str) -> Any: ...
    def unload_trik(self, trik_id: str) -> bool: ...


class GatewayRegistryProvider:
    """Implements TrikRegistryContext — provides trik management operations."""

    def __init__(
        self,
        config_dir: str,
        gateway: RegistryProviderGateway,
        registry_base_url: str = "https://api.trikhub.com",
    ) -> None:
        self._base_url = registry_base_url
        self._config_dir = Path(config_dir)
        self._gateway = gateway

    async def search(
        self, query: str, page: int = 1, page_size: int = 10
    ) -> TrikSearchResult:
        url = f"{self._base_url}/api/v1/triks"
        params = {"q": query, "page": str(page), "pageSize": str(page_size)}

        async with httpx.AsyncClient() as client:
            res = await client.get(url, params=params)
            if res.status_code != 200:
                raise RuntimeError(f"Registry search failed: {res.status_code}")

            data = res.json()
            return TrikSearchResult(
                triks=[
                    TrikSearchResultItem(
                        name=t.get("name", ""),
                        description=t.get("description", "")[:200],
                        version=t.get("latestVersion", "0.0.0"),
                        downloads=t.get("totalDownloads", 0),
                        verified=t.get("verified", False),
                    )
                    for t in data.get("triks", [])
                ],
                total=data.get("total", 0),
                hasMore=data.get("hasMore", False),
            )

    async def list(self) -> list[InstalledTrikInfo]:
        triks = self._gateway.get_loaded_triks()
        result: list[InstalledTrikInfo] = []

        for trik_id, loaded in triks.items():
            m = loaded.manifest
            caps: list[str] = []
            if m.capabilities:
                if getattr(m.capabilities.session, "enabled", False):
                    caps.append("session")
                if getattr(m.capabilities.storage, "enabled", False):
                    caps.append("storage")
                if getattr(m.capabilities.filesystem, "enabled", False):
                    caps.append("filesystem")
                if getattr(m.capabilities.shell, "enabled", False):
                    caps.append("shell")
                if getattr(m.capabilities.trikManagement, "enabled", False):
                    caps.append("trikManagement")

            result.append(
                InstalledTrikInfo(
                    id=trik_id,
                    name=getattr(m, "name", trik_id) or trik_id,
                    version=getattr(m, "version", "0.0.0") or "0.0.0",
                    mode=getattr(getattr(m, "agent", None), "mode", "tool") or "tool",
                    description=(getattr(m, "description", "") or "")[:200],
                    capabilities=caps,
                )
            )

        return result

    async def install(
        self, trik_id: str, version: str | None = None
    ) -> TrikInstallResult:
        loaded_triks = self._gateway.get_loaded_triks()
        if trik_id in loaded_triks:
            existing = loaded_triks[trik_id]
            return TrikInstallResult(
                status="already_installed",
                trikId=trik_id,
                version=getattr(existing.manifest, "version", "0.0.0") or "0.0.0",
            )

        try:
            async with httpx.AsyncClient() as client:
                res = await client.get(
                    f"{self._base_url}/api/v1/triks/{trik_id}"
                )
                if res.status_code != 200:
                    return TrikInstallResult(
                        status="failed",
                        trikId=trik_id,
                        version="",
                        error=f"Trik not found: {trik_id}",
                    )
                trik_info = res.json()

            versions = trik_info.get("versions", [])
            target = None
            if version:
                target = next(
                    (v for v in versions if v["version"] == version), None
                )
            else:
                target = next(
                    (
                        v
                        for v in versions
                        if v["version"] == trik_info.get("latestVersion")
                    ),
                    None,
                )

            if not target:
                return TrikInstallResult(
                    status="failed",
                    trikId=trik_id,
                    version=version or "",
                    error="Version not found",
                )

            trik_runtime = target.get("runtime", "python")
            is_same_runtime = trik_runtime == "python"
            project_root = self._get_project_root()
            has_pyproject = (project_root / "pyproject.toml").exists()
            needs_container = self._needs_containerization(target.get("manifest"))

            # Containerized triks (filesystem/shell) always go to .trikhub/triks/
            # so the Docker mount gets a self-contained directory.
            use_pip = is_same_runtime and has_pyproject and not needs_container

            if use_pip:
                # Same runtime, non-containerized: install via pip from git URL
                git_tag = target["gitTag"]
                pip_url = f"git+https://github.com/{trik_info['githubRepo']}@{git_tag}"
                subprocess.run(
                    [sys.executable, "-m", "pip", "install", pip_url, "--quiet"],
                    check=True,
                    capture_output=True,
                )
                # Discover installed path for loading
                trik_path = self._find_pip_installed_trik(trik_id)
            else:
                # Cross-language, containerized, or no pyproject.toml: download to .trikhub/triks/
                self._download_to_triks_dir(trik_id, trik_info["githubRepo"], target["gitTag"])
                trik_path = str(self._get_trik_dir(trik_id))

            self._add_to_config(trik_id, trik_runtime)
            await self._gateway.load_trik(trik_path)

            return TrikInstallResult(
                status="installed",
                trikId=trik_id,
                version=target["version"],
            )
        except Exception as e:
            return TrikInstallResult(
                status="failed",
                trikId=trik_id,
                version=version or "",
                error=str(e)[:200],
            )

    async def uninstall(self, trik_id: str) -> TrikUninstallResult:
        if trik_id not in self._gateway.get_loaded_triks():
            return TrikUninstallResult(status="not_found", trikId=trik_id)

        try:
            self._gateway.unload_trik(trik_id)

            # Try to remove from .trikhub/triks/ (cross-language)
            trik_dir = self._get_trik_dir(trik_id)
            if trik_dir.exists():
                shutil.rmtree(trik_dir)

            # Try to remove via pip (same-runtime Python triks)
            self._try_pip_uninstall(trik_id)

            self._remove_from_config(trik_id)

            return TrikUninstallResult(status="uninstalled", trikId=trik_id)
        except Exception as e:
            return TrikUninstallResult(
                status="failed",
                trikId=trik_id,
                error=str(e)[:200],
            )

    async def upgrade(
        self, trik_id: str, version: str | None = None
    ) -> TrikUpgradeResult:
        loaded = self._gateway.get_loaded_triks().get(trik_id)
        if not loaded:
            return TrikUpgradeResult(
                status="not_found",
                trikId=trik_id,
                previousVersion="",
                newVersion="",
            )

        previous_version = getattr(loaded.manifest, "version", "0.0.0") or "0.0.0"

        try:
            async with httpx.AsyncClient() as client:
                res = await client.get(
                    f"{self._base_url}/api/v1/triks/{trik_id}"
                )
                if res.status_code != 200:
                    return TrikUpgradeResult(
                        status="failed",
                        trikId=trik_id,
                        previousVersion=previous_version,
                        newVersion="",
                        error="Registry fetch failed",
                    )
                trik_info = res.json()

            target_version = version or trik_info.get("latestVersion", "")

            if target_version == previous_version:
                return TrikUpgradeResult(
                    status="already_latest",
                    trikId=trik_id,
                    previousVersion=previous_version,
                    newVersion=previous_version,
                )

            # Uninstall then reinstall
            await self.uninstall(trik_id)
            install_result = await self.install(trik_id, target_version)

            if install_result.status == "failed":
                return TrikUpgradeResult(
                    status="failed",
                    trikId=trik_id,
                    previousVersion=previous_version,
                    newVersion=target_version,
                    error=install_result.error,
                )

            return TrikUpgradeResult(
                status="upgraded",
                trikId=trik_id,
                previousVersion=previous_version,
                newVersion=target_version,
            )
        except Exception as e:
            return TrikUpgradeResult(
                status="failed",
                trikId=trik_id,
                previousVersion=previous_version,
                newVersion=version or "",
                error=str(e)[:200],
            )

    async def get_info(self, trik_id: str) -> TrikDetailInfo | None:
        try:
            async with httpx.AsyncClient() as client:
                res = await client.get(
                    f"{self._base_url}/api/v1/triks/{trik_id}"
                )
                if res.status_code != 200:
                    return None

                data = res.json()
                return TrikDetailInfo(
                    name=data.get("name", ""),
                    description=data.get("description", "")[:200],
                    latestVersion=data.get("latestVersion", "0.0.0"),
                    versions=[v["version"] for v in data.get("versions", [])],
                    downloads=data.get("totalDownloads", 0),
                    verified=data.get("verified", False),
                    mode=data.get("versions", [{}])[0]
                    .get("manifest", {})
                    .get("agent", {})
                    .get("mode", "tool")
                    if data.get("versions")
                    else "tool",
                )
        except Exception:
            return None

    # --------------------------------------------------------------------------
    # Private helpers
    # --------------------------------------------------------------------------

    @staticmethod
    def _needs_containerization(manifest: dict | None) -> bool:
        """Check if a manifest declares filesystem or shell capabilities."""
        if not manifest:
            return False
        caps = manifest.get("capabilities") or {}
        fs = caps.get("filesystem") or {}
        sh = caps.get("shell") or {}
        return bool(fs.get("enabled") or sh.get("enabled"))

    def _get_project_root(self) -> Path:
        """Get the project root directory (parent of configDir/.trikhub)."""
        return self._config_dir.parent

    def _download_to_triks_dir(
        self, trik_id: str, github_repo: str, git_tag: str
    ) -> None:
        """Download a cross-language trik to .trikhub/triks/."""
        trik_dir = self._get_trik_dir(trik_id)
        if trik_dir.exists():
            shutil.rmtree(trik_dir)
        trik_dir.mkdir(parents=True, exist_ok=True)

        git_url = f"https://github.com/{github_repo}.git"
        subprocess.run(
            ["git", "clone", "--depth", "1", "--branch", git_tag, git_url, str(trik_dir)],
            check=True,
            capture_output=True,
        )

        git_dir = trik_dir / ".git"
        if git_dir.exists():
            shutil.rmtree(git_dir)

        # Install dependencies for JS triks (detect package manager from lockfile)
        if (trik_dir / "package.json").exists():
            if (trik_dir / "pnpm-lock.yaml").exists():
                cmd = ["pnpm", "install", "--frozen-lockfile", "--prod"]
            elif (trik_dir / "yarn.lock").exists():
                cmd = ["yarn", "install", "--production", "--frozen-lockfile"]
            else:
                cmd = ["npm", "install", "--production"]
            subprocess.run(
                cmd,
                cwd=str(trik_dir),
                check=False,
                capture_output=True,
            )
        elif (trik_dir / "requirements.txt").exists():
            subprocess.run(
                [sys.executable, "-m", "pip", "install", "-r", "requirements.txt", "--quiet"],
                cwd=str(trik_dir),
                check=False,
                capture_output=True,
            )

        # Write identity file for trusted scoped name
        identity_path = trik_dir / ".trikhub-identity.json"
        identity = {
            "scopedName": trik_id,
            "installedAt": datetime.now(timezone.utc).isoformat(),
        }
        identity_path.write_text(json.dumps(identity, indent=2))

    def _find_pip_installed_trik(self, trik_id: str) -> str:
        """Find the installed path for a pip-installed Python trik."""
        # Try to discover via trikhub discovery if available
        try:
            from trikhub.cli.discovery import discover_triks_in_site_packages
            for t in discover_triks_in_site_packages():
                if t.package_name == trik_id or t.trik_id == trik_id:
                    return str(t.path)
        except Exception:
            pass
        # Fallback: return the trik_id and let the gateway resolve it
        return trik_id

    def _try_pip_uninstall(self, trik_id: str) -> None:
        """Try to uninstall a Python trik via pip."""
        # Generate pip package name candidates from @scope/name
        if trik_id.startswith("@"):
            parts = trik_id.lstrip("@").split("/", 1)
            if len(parts) == 2:
                candidates = [parts[1].replace("_", "-"), f"{parts[0]}-{parts[1].replace('_', '-')}"]
            else:
                candidates = [parts[0]]
        else:
            candidates = [trik_id.replace("_", "-")]

        for pip_name in candidates:
            try:
                subprocess.run(
                    [sys.executable, "-m", "pip", "uninstall", pip_name, "-y", "--quiet"],
                    check=True,
                    capture_output=True,
                )
                break
            except subprocess.CalledProcessError:
                continue

    def _get_trik_dir(self, trik_id: str) -> Path:
        if trik_id.startswith("@"):
            scope, name = trik_id.split("/", 1)
            return self._config_dir / "triks" / scope / name
        return self._config_dir / "triks" / trik_id

    def _add_to_config(self, trik_id: str, runtime: str | None = None) -> None:
        config_path = self._config_dir / "config.json"
        if config_path.exists():
            config = json.loads(config_path.read_text())
        else:
            config = {"triks": []}

        if trik_id not in config["triks"]:
            config["triks"].append(trik_id)

        if runtime:
            if "runtimes" not in config:
                config["runtimes"] = {}
            config["runtimes"][trik_id] = runtime

        config_path.write_text(json.dumps(config, indent=2))

    def _remove_from_config(self, trik_id: str) -> None:
        config_path = self._config_dir / "config.json"
        if not config_path.exists():
            return

        config = json.loads(config_path.read_text())
        config["triks"] = [t for t in config.get("triks", []) if t != trik_id]

        # Clean up version and runtime tracking
        if "trikhub" in config:
            config["trikhub"].pop(trik_id, None)
        if "runtimes" in config:
            config["runtimes"].pop(trik_id, None)

        config_path.write_text(json.dumps(config, indent=2))

File: packages/js/gateway/src/registry-provider.ts
/**
 * Gateway Registry Provider — implements TrikRegistryContext.
 *
 * Provides search, install, uninstall, upgrade, list, and getInfo operations
 * by proxying to the TrikHub registry API and the local gateway.
 *
 * Mirrors packages/python/trikhub/gateway/registry_provider.py.
 */

import { execSync } from 'node:child_process';
import { existsSync, rmSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { join, resolve } from 'node:path';
import type {
  TrikRegistryContext,
  TrikSearchResult,
  InstalledTrikInfo,
  TrikInstallResult,
  TrikUninstallResult,
  TrikUpgradeResult,
  TrikDetailInfo,
  TrikManifest,
} from '@trikhub/manifest';

export interface RegistryProviderGateway {
  getLoadedTriks(): Map<string, { manifest: TrikManifest; path: string }>;
  loadTrik(path: string): Promise<TrikManifest>;
  unloadTrik(id: string): boolean;
}

export interface RegistryProviderConfig {
  registryBaseUrl?: string;
  configDir: string;
  gateway: RegistryProviderGateway;
}

export class GatewayRegistryProvider implements TrikRegistryContext {
  private baseUrl: string;
  private configDir: string;
  private gateway: RegistryProviderGateway;

  constructor(config: RegistryProviderConfig) {
    this.baseUrl = config.registryBaseUrl ?? 'https://api.trikhub.com';
    this.configDir = config.configDir;
    this.gateway = config.gateway;
  }

  async search(
    query: string,
    options?: { page?: number; pageSize?: number }
  ): Promise<TrikSearchResult> {
    const page = options?.page ?? 1;
    const pageSize = options?.pageSize ?? 10;
    const url = `${this.baseUrl}/api/v1/triks?q=${encodeURIComponent(query)}&page=${page}&pageSize=${pageSize}`;

    const res = await fetch(url);
    if (!res.ok) {
      throw new Error(`Registry search failed: ${res.status}`);
    }

    const data: any = await res.json();
    return {
      triks: (data.triks ?? []).map((t: any) => ({
        name: t.name,
        description: (t.description ?? '').slice(0, 200),
        version: t.latestVersion ?? '0.0.0',
        downloads: t.totalDownloads ?? 0,
        verified: t.verified ?? false,
      })),
      total: data.total ?? 0,
      hasMore: data.hasMore ?? false,
    };
  }

  async list(): Promise<InstalledTrikInfo[]> {
    const triks = this.gateway.getLoadedTriks();
    const result: InstalledTrikInfo[] = [];

    for (const [id, loaded] of triks) {
      const m = loaded.manifest;
      const caps: string[] = [];
      if (m.capabilities?.session?.enabled) caps.push('session');
      if (m.capabilities?.storage?.enabled) caps.push('storage');
      if (m.capabilities?.filesystem?.enabled) caps.push('filesystem');
      if (m.capabilities?.shell?.enabled) caps.push('shell');
      if (m.capabilities?.trikManagement?.enabled) caps.push('trikManagement');

      result.push({
        id,
        name: m.name ?? id,
        version: m.version ?? '0.0.0',
        mode: m.agent?.mode ?? 'tool',
        description: (m.description ?? '').slice(0, 200),
        capabilities: caps,
      });
    }

    return result;
  }

  async install(trikId: string, version?: string): Promise<TrikInstallResult> {
    if (this.gateway.getLoadedTriks().has(trikId)) {
      const existing = this.gateway.getLoadedTriks().get(trikId)!;
      return {
        status: 'already_installed',
        trikId,
        version: existing.manifest.version ?? '0.0.0',
      };
    }

    try {
      const res = await fetch(
        `${this.baseUrl}/api/v1/triks/${encodeURIComponent(trikId)}`
      );
      if (!res.ok) {
        return {
          status: 'failed',
          trikId,
          version: '',
          error: `Trik not found: ${trikId}`,
        };
      }

      const trikInfo: any = await res.json();

      const targetVersion = version
        ? trikInfo.versions?.find((v: any) => v.version === version)
        : trikInfo.versions?.find(
            (v: any) => v.version === trikInfo.latestVersion
          );

      if (!targetVersion) {
        return {
          status: 'failed',
          trikId,
          version: version ?? '',
          error: 'Version not found',
        };
      }

      const trikRuntime = targetVersion.runtime ?? 'node';
      const isSameRuntime = trikRuntime === 'node';
      const projectRoot = this.getProjectRoot();
      const hasPackageJson = existsSync(join(projectRoot, 'package.json'));
      const needsContainer = this.needsContainerization(targetVersion.manifest);

      // Containerized triks (filesystem/shell) always go to .trikhub/triks/
      // so the Docker mount gets a self-contained directory with node_modules.
      // Non-containerized same-runtime triks use package.json + node_modules.
      const usePackageManager = isSameRuntime && hasPackageJson && !needsContainer;

      if (usePackageManager) {
        await this.installViaPackageManager(trikId, trikInfo.githubRepo, targetVersion.gitTag);
      } else {
        this.downloadToTriksDir(trikId, trikInfo.githubRepo, targetVersion.gitTag);
      }

      this.addToConfig(trikId, trikRuntime);
      await this.gateway.loadTrik(
        usePackageManager
          ? join(projectRoot, 'node_modules', ...trikId.split('/'))
          : this.getTrikDir(trikId)
      );

      return {
        status: 'installed',
        trikId,
        version: targetVersion.version,
      };
    } catch (err: any) {
      return {
        status: 'failed',
        trikId,
        version: version ?? '',
        error: (err.message ?? 'Unknown error').slice(0, 200),
      };
    }
  }

  async uninstall(trikId: string): Promise<TrikUninstallResult> {
    if (!this.gateway.getLoadedTriks().has(trikId)) {
      return { status: 'not_found', trikId };
    }

    try {
      this.gateway.unloadTrik(trikId);

      // Check if trik lives in node_modules (same-runtime) or .trikhub/triks/ (cross-language)
      const projectRoot = this.getProjectRoot();
      const nmPath = join(projectRoot, 'node_modules', ...trikId.split('/'));
      const trikDir = this.getTrikDir(trikId);

      if (existsSync(nmPath)) {
        // Same-runtime: remove from package.json and node_modules
        rmSync(nmPath, { recursive: true, force: true });
        this.removeFromPackageJson(trikId);
      }

      if (existsSync(trikDir)) {
        // Cross-language: remove from .trikhub/triks/
        rmSync(trikDir, { recursive: true, force: true });
      }

      this.removeFromConfig(trikId);

      return { status: 'uninstalled', trikId };
    } catch (err: any) {
      return {
        status: 'failed',
        trikId,
        error: (err.message ?? 'Unknown error').slice(0, 200),
      };
    }
  }

  async upgrade(
    trikId: string,
    version?: string
  ): Promise<TrikUpgradeResult> {
    const loaded = this.gateway.getLoadedTriks().get(trikId);
    if (!loaded) {
      return {
        status: 'not_found',
        trikId,
        previousVersion: '',
        newVersion: '',
      };
    }

    const previousVersion = loaded.manifest.version ?? '0.0.0';

    try {
      const res = await fetch(
        `${this.baseUrl}/api/v1/triks/${encodeURIComponent(trikId)}`
      );
      if (!res.ok) {
        return {
          status: 'failed',
          trikId,
          previousVersion,
          newVersion: '',
          error: 'Registry fetch failed',
        };
      }

      const trikInfo: any = await res.json();
      const targetVersion = version ?? trikInfo.latestVersion;

      if (targetVersion === previousVersion) {
        return {
          status: 'already_latest',
          trikId,
          previousVersion,
          newVersion: previousVersion,
        };
      }

      // Uninstall then reinstall
      await this.uninstall(trikId);
      const installResult = await this.install(trikId, targetVersion);

      if (installResult.status === 'failed') {
        return {
          status: 'failed',
          trikId,
          previousVersion,
          newVersion: targetVersion,
          error: installResult.error,
        };
      }

      return {
        status: 'upgraded',
        trikId,
        previousVersion,
        newVersion: targetVersion,
      };
    } catch (err: any) {
      return {
        status: 'failed',
        trikId,
        previousVersion,
        newVersion: version ?? '',
        error: (err.message ?? 'Unknown error').slice(0, 200),
      };
    }
  }

  async getInfo(trikId: string): Promise<TrikDetailInfo | null> {
    try {
      const res = await fetch(
        `${this.baseUrl}/api/v1/triks/${encodeURIComponent(trikId)}`
      );
      if (!res.ok) return null;

      const data: any = await res.json();
      return {
        name: data.name,
        description: (data.description ?? '').slice(0, 200),
        latestVersion: data.latestVersion ?? '0.0.0',
        versions: (data.versions ?? []).map((v: any) => v.version),
        downloads: data.totalDownloads ?? 0,
        verified: data.verified ?? false,
        mode: data.versions?.[0]?.manifest?.agent?.mode ?? 'tool',
      };
    } catch {
      return null;
    }
  }

  // --------------------------------------------------------------------------
  // Private helpers
  // --------------------------------------------------------------------------

  /**
   * Install a same-runtime JS trik via git URL in package.json.
   * Mirrors the CLI's installFromTrikhub behavior.
   */
  private async installViaPackageManager(
    trikId: string,
    githubRepo: string,
    gitTag: string
  ): Promise<void> {
    const projectRoot = this.getProjectRoot();
    const packageJsonPath = join(projectRoot, 'package.json');
    const pkg = JSON.parse(readFileSync(packageJsonPath, 'utf-8'));

    if (!pkg.dependencies) {
      pkg.dependencies = {};
    }
    pkg.dependencies[trikId] = `github:${githubRepo}#${gitTag}`;
    writeFileSync(packageJsonPath, JSON.stringify(pkg, null, 2) + '\n');

    // Remove existing from node_modules to force fresh install
    const nmPath = join(projectRoot, 'node_modules', ...trikId.split('/'));
    if (existsSync(nmPath)) {
      rmSync(nmPath, { recursive: true, force: true });
    }

    const pm = this.detectPackageManager();
    const installCmd = pm === 'pnpm'
      ? `pnpm install`
      : pm === 'yarn'
        ? `yarn install`
        : `npm install`;
    execSync(installCmd, { cwd: projectRoot, stdio: 'pipe' });
  }

  /**
   * Download a cross-language trik to .trikhub/triks/.
   */
  private downloadToTriksDir(
    trikId: string,
    githubRepo: string,
    gitTag: string
  ): void {
    const trikDir = this.getTrikDir(trikId);
    mkdirSync(trikDir, { recursive: true });

    const gitUrl = `https://github.com/${githubRepo}.git`;
    execSync(
      `git clone --depth 1 --branch ${gitTag} ${gitUrl} ${trikDir}`,
      { stdio: 'pipe' }
    );

    // Remove .git to save space
    const gitDir = join(trikDir, '.git');
    if (existsSync(gitDir)) {
      rmSync(gitDir, { recursive: true, force: true });
    }

    // Install dependencies
    if (existsSync(join(trikDir, 'package.json'))) {
      const cmd = existsSync(join(trikDir, 'pnpm-lock.yaml'))
        ? 'pnpm install --frozen-lockfile --prod'
        : existsSync(join(trikDir, 'yarn.lock'))
          ? 'yarn install --production --frozen-lockfile'
          : 'npm install --production';
      execSync(cmd, { cwd: trikDir, stdio: 'pipe' });
    } else if (existsSync(join(trikDir, 'requirements.txt'))) {
      execSync('pip install -r requirements.txt', { cwd: trikDir, stdio: 'pipe' });
    }

    // Write identity file for trusted scoped name
    const identityPath = join(trikDir, '.trikhub-identity.json');
    const identity = {
      scopedName: trikId,
      installedAt: new Date().toISOString(),
    };
    writeFileSync(identityPath, JSON.stringify(identity, null, 2));
  }

  /**
   * Check if a manifest declares filesystem or shell capabilities,
   * which require containerized (Docker) execution.
   */
  private needsContainerization(manifest?: any): boolean {
    const caps = manifest?.capabilities;
    return !!(caps?.filesystem?.enabled || caps?.shell?.enabled);
  }

  /**
   * Get the project root directory (parent of configDir/.trikhub).
   */
  private getProjectRoot(): string {
    return resolve(this.configDir, '..');
  }

  /**
   * Detect which package manager is used in the project.
   */
  private detectPackageManager(): 'npm' | 'pnpm' | 'yarn' {
    const root = this.getProjectRoot();
    if (existsSync(join(root, 'pnpm-lock.yaml'))) return 'pnpm';
    if (existsSync(join(root, 'yarn.lock'))) return 'yarn';
    return 'npm';
  }

  private getTrikDir(trikId: string): string {
    if (trikId.startsWith('@')) {
      const [scope, name] = trikId.split('/');
      return join(this.configDir, 'triks', scope, name);
    }
    return join(this.configDir, 'triks', trikId);
  }

  private addToConfig(trikId: string, runtime?: string): void {
    const configPath = join(this.configDir, 'config.json');
    const config = existsSync(configPath)
      ? JSON.parse(readFileSync(configPath, 'utf-8'))
      : { triks: [] };

    if (!config.triks.includes(trikId)) {
      config.triks.push(trikId);
    }

    if (runtime) {
      if (!config.runtimes) config.runtimes = {};
      config.runtimes[trikId] = runtime;
    }

    writeFileSync(configPath, JSON.stringify(config, null, 2));
  }

  private removeFromPackageJson(trikId: string): void {
    const packageJsonPath = join(this.getProjectRoot(), 'package.json');
    if (!existsSync(packageJsonPath)) return;

    try {
      const pkg = JSON.parse(readFileSync(packageJsonPath, 'utf-8'));
      if (pkg.dependencies?.[trikId]) {
        delete pkg.dependencies[trikId];
        writeFileSync(packageJsonPath, JSON.stringify(pkg, null, 2) + '\n');
      }
    } catch {
      // package.json parsing failed — skip
    }
  }

  private removeFromConfig(trikId: string): void {
    const configPath = join(this.configDir, 'config.json');
    if (!existsSync(configPath)) return;

    const config = JSON.parse(readFileSync(configPath, 'utf-8'));
    config.triks = (config.triks ?? []).filter((t: string) => t !== trikId);

    // Clean up version and runtime tracking
    if (config.trikhub) {
      delete config.trikhub[trikId];
    }
    if (config.runtimes) {
      delete config.runtimes[trikId];
    }

    writeFileSync(configPath, JSON.stringify(config, null, 2));
  }
}
